atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/transaction/WitnessedTransactionsBuilder.java

```diff
@@ -0,0 +1,92 @@
+
+import com.palantir.atlasdb.workload.store.ImmutableWorkloadCell;
+import com.palantir.atlasdb.workload.transaction.witnessed.ImmutableWitnessedTransaction;
+import com.palantir.atlasdb.workload.transaction.witnessed.WitnessedDeleteTransactionAction;
+import com.palantir.atlasdb.workload.transaction.witnessed.WitnessedReadTransactionAction;
+import com.palantir.atlasdb.workload.transaction.witnessed.WitnessedTransaction;
+import com.palantir.atlasdb.workload.transaction.witnessed.WitnessedTransactionAction;
+import com.palantir.atlasdb.workload.transaction.witnessed.WitnessedWriteTransactionAction;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicLong;
+
+public final class WitnessedTransactionsBuilder {
+
+    private final List<WitnessedTransaction> witnessedTransactions = new ArrayList<>();
+
+    private final AtomicLong timestampCounter = new AtomicLong();
+    private final String table;
+
+    public WitnessedTransactionsBuilder(String table) {
+        this.table = table;
+    }
+
+    public WitnessedTransactionBuilder startTransaction() {
+        return new WitnessedTransactionBuilder();
+    }
+
+    public List<WitnessedTransaction> build() {
+        return List.copyOf(witnessedTransactions);
+    }
+
+    public final class WitnessedTransactionBuilder {
+        private final List<WitnessedTransactionAction> actions = new ArrayList<>();
+
+        private boolean needsCommitTimestamp = false;
+
+        public WitnessedTransactionBuilder read(Integer row, Integer column, Optional<Integer> valueRead) {
+            actions.add(WitnessedReadTransactionAction.of(table, ImmutableWorkloadCell.of(row, column), valueRead));
+            return this;
+        }
+
+        public WitnessedTransactionBuilder read(Integer row, Integer column, Integer valueRead) {
+            return read(row, column, Optional.of(valueRead));
+        }
+
+        public WitnessedTransactionBuilder read(Integer row, Integer column) {
+            return read(row, column, Optional.empty());
+        }
+
+        public WitnessedTransactionBuilder write(Integer row, Integer column, Integer value) {
+            actions.add(WitnessedWriteTransactionAction.of(table, ImmutableWorkloadCell.of(row, column), value));
+            needsCommitTimestamp = true;
+            return this;
+        }
+
+        public WitnessedTransactionBuilder delete(Integer row, Integer column) {
+            actions.add(WitnessedDeleteTransactionAction.of(table, ImmutableWorkloadCell.of(row, column)));
+            needsCommitTimestamp = true;
+            return this;
+        }
+
+        public WitnessedTransactionsBuilder endTransaction() {
+            long startTimestamp = timestampCounter.incrementAndGet();
+            Optional<Long> commitTimestamp =
+                    needsCommitTimestamp ? Optional.of(timestampCounter.incrementAndGet()) : Optional.empty();
```

Comment:
This may not be the intended usage, but the semantics of something like this wouldn't match what I'd expect:
```
WitnessedTransactionsBuilder transactionsBuilder = new WitnessedTransactionsBuilder(table);
WitnessedTransactionBuilder txn1 = transactionsBuilder.startTransaction().write(1,2,3);
WitnessedTransactionBuilder txn2 = transactionsBuilder.startTransaction().read(1,2);
txn2.endTransaction();
return txn1.endTransaction().build(); // this has [txn2, txn1] with no overlap - but I'd expect either just [txn1] or [txn2, txn1] where txn1 started before txn2!
```

I think this makes sense for what you're testing and how you're using, though probably want to have some comments somewhere indicating that witnessed transaction objects created are only "correct"/representative within the scope of a single building flow (and probably some usage instructions) - otherwise this might cause issues when we decide to try and reuse this in a different context in the future.

(Alternatively, if the intention was for this to be a persistent data structure then you can use a Vavr rather than Java list, and have the thing returned by endTransaction create a new WitnessedTransactionsBuilder with the new list.)

Reply:
Good catch! Yeah, I think the solution here is that we should capture the `startTimestamp` as soon as we start the transaction. We can really go the extra mile here, and actually allow clients to specify an optional commit timestamp, which makes this type of ordering a lot easier!

Reply:
That way you get the behavior you mentioned, which I think is desirable!